llvm/tools/sycl-post-link/SYCLStripDeadDebugInfo.cpp

```diff
@@ -0,0 +1,123 @@
+//===-- SYCLStripDeadDebugInfo.cpp - Strip debug info from split module ---===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SYCLStripDeadDebugInfo.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/DebugInfo.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/Module.h"
+
+#include <set>
+#include <unordered_set>
+
+using namespace llvm;
+
+namespace {
+
+// The implementation of this function is almost the same as of the one for
+// llvm::StripDeadDebugInfoPass, but some conditions for delete debug info more
+// aggressive taking into account that processed module is split from another
+// bigger module and contains only a subset of functions from it.
+bool stripDeadDebugInfoImpl(Module &M) {
+  bool Changed = false;
+
+  LLVMContext &C = M.getContext();
+
+  // Find all debug info in DIF. This is actually overkill in terms of what we
+  // want to do, but we want to try and be as resilient as possible in the face
+  // of potential debug info changes by using the formal interfaces given to us
+  // as much as possible.
+  DebugInfoFinder DIF;
+  DIF.processModule(M);
+
+  // For each compile unit, find the live set of global variables/functions and
+  // replace the current list of potentially dead global variables/functions
+  // with the live list.
+  SmallVector<Metadata *, 64> LiveGlobalVariables;
+  DenseSet<DIGlobalVariableExpression *> VisitedSet;
+
+  std::set<DIGlobalVariableExpression *> LiveGVs;
+  for (GlobalVariable &GV : M.globals()) {
+    SmallVector<DIGlobalVariableExpression *, 1> GVEs;
+    GV.getDebugInfo(GVEs);
+    for (auto *GVE : GVEs)
+      LiveGVs.insert(GVE);
+  }
+
+  std::unordered_set<const DISubprogram *> PresentFunctionsSPs;
```

Comment:
It took me some time to figure out the exact change compared to the original version of the algorithm, could you please add an explicit comment highlighting this change?

My understanding is that the change is to expect some `DISubprogram` not being associated with any functions (due to functions being removed by device code split earlier) and therefore, not to mark corresponding compilation units as live, right?

Reply:
Comments were added.
You are right. First optimization is to skip compile units if they have no `DISubprogram`s associated with any function in current submodule.
Second proposed optimization is to skip compile units with constant global expressions that also have no `DISubprogram`s associated with any function in current submodule.